### PR TITLE
Support Windows

### LINKED ISSUE #3: Problems with installing on Windows
When trying to pip install the package on Windows I get the following error
```
(vevn) C:\Users\camo\local\src\ap_features>python -m pip install -e .
Obtaining file:///C:/Users/camo/local/src/ap_features
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: numba in c:\users\camo\local\src\ap_features\vevn\lib\site-packages (from ap-features==0.1.2) (0.53.1)
Requirement already satisfied: tqdm in c:\users\camo\local\src\ap_features\vevn\lib\site-packages (from ap-features==0.1.2) (4.61.0)
Requirement already satisfied: numpy in c:\users\camo\local\src\ap_features\vevn\lib\site-packages (from ap-features==0.1.2) (1.20.3)
Requirement already satisfied: setuptools in c:\users\camo\local\src\ap_features\vevn\lib\site-packages (from numba->ap-features==0.1.2) (57.0.0)
Requirement already satisfied: llvmlite<0.37,>=0.36.0rc1 in c:\users\camo\local\src\ap_features\vevn\lib\site-packages (from numba->ap-features==0.1.2) (0.36.0)
Installing collected packages: ap-features
  Running setup.py develop for ap-features
    ERROR: Command errored out with exit status 1:
     command: 'C:\Users\camo\local\src\ap_features\vevn\Scripts\python.exe' -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\camo\\local\\src\\ap_features\\setup.py'"'"'; __file__='"'"'C:\\Users\\camo\\local\\src\\ap_features\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
         cwd: C:\Users\camo\local\src\ap_features\
    Complete output (76 lines):
    running develop
    running egg_info
    writing ap_features.egg-info\PKG-INFO
    writing dependency_links to ap_features.egg-info\dependency_links.txt
    writing requirements to ap_features.egg-info\requires.txt
    writing top-level names to ap_features.egg-info\top_level.txt
    reading manifest file 'ap_features.egg-info\SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no previously-included files matching '__pycache__' found under directory '*'
    warning: no previously-included files matching '*.py[co]' found under directory '*'
    warning: no files found matching '*.jpg' under directory 'docs'
    warning: no files found matching '*.png' under directory 'docs'
    warning: no files found matching '*.gif' under directory 'docs'
    warning: no files found matching '*.md' under directory 'docs'
    warning: no directories found matching 'build'
    adding license file 'LICENSE'
    writing manifest file 'ap_features.egg-info\SOURCES.txt'
    running build_ext
    building 'ap_features.cost_terms' extension
    C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30037\bin\HostX86\x86\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -IC:\Users\camo\local\src\ap_features\vevn\include -IC:\Users\camo\AppData\Local\Programs\Python\Python39\include -IC:\Users\camo\AppData\Local\Programs\Python\Python39\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30037\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt -IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30037\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt /Tcsrc/cost_terms.c /Fobuild\temp.win32-3.9\Release\src/cost_terms.obj
    cost_terms.c
    src/cost_terms.c(23): error C2059: syntax error: '<parameter-list>'
    src/cost_terms.c(53): error C2059: syntax error: '<parameter-list>'
    src/cost_terms.c(204): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(204): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(208): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(208): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(210): warning C4047: '<': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(210): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(204): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(208): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(210): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(226): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(226): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(230): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(230): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(232): warning C4047: '<': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(232): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(226): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(230): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(232): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(247): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(247): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(253): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(253): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(257): warning C4047: '<': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(257): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(247): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(253): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(257): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(272): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(272): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(278): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(278): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(282): warning C4047: '<': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(282): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(272): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(278): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(282): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(312): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(312): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(316): warning C4047: '<': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(316): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(312): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(316): error C2440: 'initializing': cannot convert from 'double *' to 'double'
    src/cost_terms.c(338): warning C4047: '>': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(338): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(338): error C2440: '=': cannot convert from 'double *' to 'double'
    src/cost_terms.c(339): warning C4047: '<': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(339): warning C4047: ':': 'double *' differs in levels of indirection from 'int'
    src/cost_terms.c(339): error C2440: '=': cannot convert from 'double *' to 'double'
    src/cost_terms.c(383): error C2057: expected constant expression
    src/cost_terms.c(383): error C2466: cannot allocate an array of constant size 0
    src/cost_terms.c(383): error C2133: 'parameter_sets_computed_per_thread': unknown size
    src/cost_terms.c(385): warning C4034: sizeof returns 0
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30037\\bin\\HostX86\\x86\\cl.exe' failed with exit code 2
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'C:\Users\camo\local\src\ap_features\vevn\Scripts\python.exe' -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\camo\\local\\src\\ap_features\\setup.py'"'"'; __file__='"'"'C:\\Users\\camo\\local\\src\\ap_features\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
```
Any ideas @KGHustad ?

### PR DESCRIPTION
**Pull request by KGHustad**
I fixed a few parts of the C code that the Microsoft C compiler did not like.

I had to change quite a few things to make sure the C functions were properly exported. The tests are completing successfully for me. I haven't really tested whether OpenMP is working properly (in the sense that things are running in parellel), but it did work when I built the C main program.

Closes #3

### CODE REVIEW

**Comment by KGHustad** [comment]
Perhaps we should merge #2 first?

**Comment by KGHustad** [comment]
#2 is merged into master now, and I have merged origin/master into this branch because it required some manual edits.

File: setup.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def check_for_openmp():
     #include <stdio.h>
 
     int main(void) {
-        #pragma omp parallel
+        #pragma omp parallel num_threads(2)
```
Comment (finsberg) [inline setup.py:22]: What does this do?

File: setup.py [hunk 2]
```diff
@@ -86,28 +86,62 @@ def get_openmp_compile_args():
...
+    elif sys.platform == "win32":
+        extra_link_args = []
     return extra_link_args
 
 
+
+def collect_c_functions(source_files):
+    import re
+
+    cdecl_re = re.compile(r"(\S+)\s+CDECL\s+(\w+)")
```
Comment (finsberg) [inline setup.py:107]: Some black magic going on here.

File: setup.py [hunk 3]
```diff
@@ -86,28 +86,62 @@ def get_openmp_compile_args():
...
+    return all_function_names
+
+
 def get_extension():
+    import os
+
     extra_compile_args = []
     extra_link_args = []
-    if check_for_openmp():
+    disable_openmp = 'DISABLE_OPENMP' in os.environ
```
Comment (finsberg) [inline setup.py:131]: This is nice!

File: src/c/cost_terms.c [hunk 4]
```diff
@@ -15,12 +15,12 @@
     (1 << 10) // number of parameter sets to compute per thread between each progress update
 
 
-int get_num_cost_terms()
+int CDECL get_num_cost_terms()
```
Comment (finsberg) [inline src/c/cost_terms.c:18]: OK

File: src/c/cost_terms.c [hunk 5]
```diff
@@ -393,31 +394,29 @@ void fill_cost_array_with_inf(double *R)
...
 #if defined(_OPENMP)
-        int thread_id = omp_get_thread_num();
+        const int thread_id = omp_get_thread_num();
 #else
-        int thread_id = 0;
+        const int thread_id = 0;
 #endif
-
-
+        // declare loop variable outside loop to make the Microsoft compiler happy
```
Comment (finsberg) [inline src/c/cost_terms.c:416]: Better keep the Microsoft compiler happy :P

File: src/c/cost_terms.c [hunk 6]
```diff
@@ -443,4 +442,11 @@ void all_cost_terms(double *R, double *traces, double *t, uint8_t *mask, long le
     // perform a final unconditional progress update
     // if the number of parameter sets is low, this will be the only progress update.
     progress_update(num_parameter_sets);
+    free(parameter_sets_computed_per_thread);
+}
+
+#ifdef _OS_WINDOWS
+void PyInit_cost_terms()
+{
```
Comment (finsberg) [inline src/c/cost_terms.c:450]: OK

**Review by finsberg — APPROVED** [review]
I must say that I am more than impressed that you managed to get this working! Really good job! I did not image that you had to all of these things to make it work. Perhaps it is easier if use C++ and pybind11 (another time).

File: setup.py (hunk 1 above)
Comment (KGHustad) [reply setup.py:22]: It makes the parallel block spawn two threads regardless of the number of cores. Without it, OpenMP would not be enabled if there is only one logical processor available or `OMP_NUM_THREADS=1` has been set in the environment.
On Windows, prior to this change the code would print one line on my quad-core CPU and detect OpenMP as working even though it wasn't.

File: setup.py (hunk 2 above)
Comment (KGHustad) [reply setup.py:107]: Indeed. It's quite easy to collect all of the function names because CDECL occurs immediately prior to the function name.

**Pull request merged**